### PR TITLE
Identity don't dispatch Analytics push event on privacy change

### PR DESCRIPTION
**Pull request by kevinlind**
## Description

Updates the Identity extension to not dispatch the `AnalyticsForIdentityRequest` event when privacy status is changed to optedout.

The Identity extension is expected to dispatch the `AnalyticsForIdentityRequest` event when the push identifier changes or when push tracking is denied the first time. This event tells Analytics to dispatch an ADBINTERNAL:Push event with `a.push.optin` value of either `True` or `False`. The issue is in some cases changing the privacy status to `optedout` would trigger a dispatch of the `AnalyticsForIdentityRequest` event even though `setPushIdentifier` was never called.

This fix changes how Identity handles processing of an privacy status change to `optedout` by clearing the push token and flags instead of calling `updatePushIdentifier` with nil, which would dispatch the event to Analytics.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/identity/src/test/java/com/adobe/marketing/mobile/identity/IdentityFunctionalTests.kt [hunk 1]
```diff
@@ -2037,4 +2042,178 @@ class IdentityFunctionalTests {
...
+            ).build()
+        )
+
+        countDownLatch.await()
+
+        assertNull(persistedData["ADOBEMOBILE_PUSH_IDENTIFIER"] as? String)
+        assertTrue(persistedData["ADOBEMOBILE_ANALYTICS_PUSH_SYNC"] as? Boolean ?: false)
+        assertFalse(persistedData["ADOBEMOBILE_PUSH_ENABLED"] as? Boolean ?: true)
+    }
+    @Test(timeout = 10000)
```
Comment (praveek) [inline code/identity/src/test/java/com/adobe/marketing/mobile/identity/IdentityFunctionalTests.kt:2145]: Can you remove the timeout param? The test does not wait on any asynchronous operation.

File: code/identity/src/test/java/com/adobe/marketing/mobile/identity/IdentityFunctionalTests.kt [hunk 2]
```diff
@@ -2037,4 +2042,178 @@ class IdentityFunctionalTests {
...
+                "com.adobe.eventSource.responseContent"
+            ).setEventData(
+                mapOf("global.privacy" to MobilePrivacyStatus.OPT_OUT.value)
+            ).build()
+        )
+
+        assertNull(identityExtension.mid)
+    }
+
+    @Test(timeout = 10000)
```
Comment (praveek) [inline code/identity/src/test/java/com/adobe/marketing/mobile/identity/IdentityFunctionalTests.kt:2176]: Same as above

**Review by praveek — APPROVED** [review]

**Review by addb — COMMENTED** [review]
Changes look good. Let's add unit tests if it is feasible to check resetIdentifier is called and does reset.

File: code/identity/src/main/java/com/adobe/marketing/mobile/identity/IdentityExtension.java [hunk 3]
```diff
@@ -499,18 +499,8 @@ void handleIdentityRequestReset(@NonNull final Event event) {
...
-        customerIds = null;
-        pushIdentifier = null;
-
-        if (namedCollection != null) {
-            namedCollection.remove(DataStoreKeys.AID_SYNCED_KEY);
-            namedCollection.remove(DataStoreKeys.PUSH_ENABLED);
-        }
 
+        clearIdentifiers();
         savePersistently(); // clear datastore
```
Comment (addb) [inline code/identity/src/main/java/com/adobe/marketing/mobile/identity/IdentityExtension.java:504]: Should we move savePersistently() to clearIdentifiers() as it is part of syncing cleared values to persistence?

File: code/identity/src/main/java/com/adobe/marketing/mobile/identity/IdentityExtension.java [hunk 4]
```diff
@@ -1857,17 +1847,7 @@ void processPrivacyChange(final Event event, final Map<String, Object> eventData
...
-            locationHint = null;
-            customerIds = null;
-
-            if (namedCollection != null) {
-                namedCollection.remove(DataStoreKeys.AID_SYNCED_KEY);
-            }
-
-            updatePushIdentifier(null);
+            clearIdentifiers();
             savePersistently(); // clear datastore
```
Comment (addb) [inline code/identity/src/main/java/com/adobe/marketing/mobile/identity/IdentityExtension.java:1851]: Should we move savePersistently() to clearIdentifiers() as it is part of syncing cleared values to persistence?

**Comment by addb** [comment]
Do we need `pushEnabled` member field L83 if we are using `isPushEnabled()` method when needed?

**Comment by addb** [comment]
> Do we need `pushEnabled` member field L83 if we are using `isPushEnabled()` method when needed?

Do you think we should add `pushEnabled`, `analyticsSynced` as member variable which we load from persistence like other fields like mid and clear similar to other fields on `savePersistently()`. Write now all the keys which are stored on persistence are member variables except for `pushEnabled` and `analyticsSynced`.

**Review by addb — APPROVED** [review]

File: code/identity/src/main/java/com/adobe/marketing/mobile/identity/IdentityExtension.java (hunk 4 above)
Comment (addb) [reply code/identity/src/main/java/com/adobe/marketing/mobile/identity/IdentityExtension.java:1848]: @kevinlind What do you think about this change?

**Comment by kevinlind** [comment]
> > Do we need `pushEnabled` member field L83 if we are using `isPushEnabled()` method when needed?
> 
> Do you think we should add `pushEnabled`, `analyticsSynced` as member variable which we load from persistence like other fields like mid and clear similar to other fields on `savePersistently()`. Write now all the keys which are stored on persistence are member variables except for `pushEnabled` and `analyticsSynced`.

This is similar to how we handle the pushEnabled and analyticsSynced flags in iOS, so I'd like to keep it this way for parity.

**Pull request merged**